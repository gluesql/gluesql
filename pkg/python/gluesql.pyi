"""
Type hints for Native Rust Extension
"""

from abc import ABCMeta
from typing import Any, Final

class Glue:
    def __init__(self, storage: "Storage") -> None: ...
    def query(sql: str): ...

class Storage(metaclass=ABCMeta):
    pass

class MemoryStorage(Storage):
    def __init__(): ...

class SharedMemoryStorage(Storage):
    def __init__(): ...

class JsonStorage(Storage):
    def __init__(path: str): ...

class SledStorageModeConfig:
    """
    In this mode, the database will make
    decisions that favor using less space
    instead of supporting the highest possible
    write throughput. This mode will also
    rewrite data more frequently as it
    strives to reduce fragmentation.
    """
    LowSpace: Final[Any]

    """
    In this mode, the database will try
    to maximize write throughput while
    potentially using more disk space.
    """
    HighThroughput: Final[Any]

class SledStorageConfig:
    def __init__(
        *,
        path: str,
        temporary: bool,
        use_compression: bool,
        print_profile_on_drop: bool,
        compression_factor: int,
        create_new: bool,
        cache_capacity: int,
        mode: "SledStorageModeConfig",
    ): ...
    path: str
    temporary: bool
    use_compression: bool
    print_profile_on_drop: bool
    compression_factor: int
    create_new: bool
    cache_capacity: int
    mode: "SledStorageModeConfig"

class SledStorage(Storage):
    def __init__(path: str): ...
    @staticmethod
    def try_from(config: "SledStorageConfig") -> "SledStorage": ...

class GlueSQLError(Exception):
    pass
