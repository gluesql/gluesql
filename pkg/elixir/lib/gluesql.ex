defmodule Gluesql do
  @moduledoc """
  Documentation for `Gluesql`.
  """

  @doc """
  Hello world.

  ## Examples

      iex> Gluesql.hello()
      :world

  """
  def hello do
    :world
  end
end
