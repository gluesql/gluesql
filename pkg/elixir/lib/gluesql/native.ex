defmodule GlueSQL.Native do
  use Rustler,
    otp_app: :gluesql,
    crate: :gluesql

  def glue_new(_storage), do: err()
  def glue_query(_glue_db, _sql), do: err()
  def memory_storage_new(), do: err()

  defp err, do: :erlang.nif_error(:nif_not_loaded)
end
