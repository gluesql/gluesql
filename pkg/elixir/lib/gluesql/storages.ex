defmodule GlueSQL.Storages do
  @moduledoc """
  To properly create SQL database with `GlueSQL.glue_new()`, the user 
  should create storage instance first.  

  `GlueSQL.Storages` module has a list of usable storages.
  """
end
