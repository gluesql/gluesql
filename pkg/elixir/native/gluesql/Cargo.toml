[package]
name = "gluesql-ex"
authors = [
  "Hoon Wee <mrgravity817@gmail.com>",
  "Taehoon Moon <taehoon.moon@outlook.com>",
]
version.workspace = true
edition.workspace = true
description.workspace = true
license.workspace = true
repository.workspace = true
documentation.workspace = true

[lib]
name = "gluesql"
path = "src/lib.rs"
crate-type = ["cdylib"]


[dependencies]
gluesql-core.workspace = true
memory-storage.workspace = true
rustler = "0.30.0"
serde_json = "1"
tokio = { version = "1", features = [
  "rt",
  "rt-multi-thread",
  "macros",
  "sync",
] }

[dev-dependencies]
test-suite.workspace = true
