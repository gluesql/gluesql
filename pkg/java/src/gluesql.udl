namespace gluesql {};

[Error]
enum GlueSQLError {
    "StorageError",
    "ExecuteError", 
    "TranslateError",
    "PlanError",
    "ParseError",
    "ValueError"
};

dictionary SledConfig {
    string? path;
    string? mode;
};

[Enum]
interface Storage {
    Memory();
    Json(string path);
    Sled(string path, SledConfig? config);
    SharedMemory(string namespace);
};

interface Glue {
    [Throws=GlueSQLError]
    constructor(Storage storage);
    [Throws=GlueSQLError]
    sequence<string> query(string sql);
};