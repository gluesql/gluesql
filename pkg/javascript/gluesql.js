import init, { Glue } from './dist/web/gluesql_js.js';

let loaded = false;

async function load() {
  await init();

  loaded = true;
}

export async function gluesql() {
  if (!loaded) {
    await load();
  }

  return new Glue();
}
