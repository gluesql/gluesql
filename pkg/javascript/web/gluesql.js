import wasm from "./Cargo.toml";

let Glue;

export async function gluesql() {
  if (!Glue) {
    Glue = (await wasm()).Glue;
  }

  return new Glue();
}
