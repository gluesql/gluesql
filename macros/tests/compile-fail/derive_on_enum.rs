use gluesql::FromGlueRow;

#[derive(FromGlueRow)]
enum E {
    A,
}

