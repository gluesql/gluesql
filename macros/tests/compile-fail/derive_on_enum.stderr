error: FromGlueRow can only be derived for structs
 --> tests/compile-fail/derive_on_enum.rs:4:1
  |
4 | enum E {
  | ^^^^

error[E0601]: `main` function not found in crate `$CRATE`
 --> tests/compile-fail/derive_on_enum.rs:6:2
  |
6 | }
  |  ^ consider adding a `main` function to `$DIR/tests/compile-fail/derive_on_enum.rs`
